File: backend/internal/services/book_service.go
package services

import (
	"book-ddd/internal/models"
	"book-ddd/internal/repositories"
)

// BookService defines the interface for book business logic.
type BookService interface {
	GetBooks() ([]models.Book, error)
	GetBook(id uint) (models.Book, error)
	CreateBook(book models.Book) (models.Book, error)
	UpdateBook(book models.Book) (models.Book, error)
	DeleteBook(id uint) error
}

type bookService struct {
	repo repositories.BookRepository
}

// NewBookService creates a new instance of the book service.
func NewBookService(r repositories.BookRepository) BookService {
	return &bookService{r}
}

func (s *bookService) GetBooks() ([]models.Book, error) {
	return s.repo.FindAll()
}

func (s *bookService) GetBook(id uint) (models.Book, error) {
	return s.repo.FindByID(id)
}

func (s *bookService) CreateBook(book models.Book) (models.Book, error) {
	return s.repo.Create(book)
}

func (s *bookService) UpdateBook(book models.Book) (models.Book, error) {
	return s.repo.Update(book)
}

func (s *bookService) DeleteBook(id uint) error {
	return s.repo.Delete(id)
}

File: src/api/BookService.js
import axios from 'axios';

const API_URL = 'http://localhost:8080/books';

const getBooks = (page = 1, limit = 10) => {
    return axios.get(API_URL, {
        params: {
            page,
            limit
        }
    });
};

const getBook = (id) => {
    return axios.get(`${API_URL}/${id}`);
};

const createBook = (book) => {
    return axios.post(API_URL, book);
};

const updateBook = (id, book) => {
    return axios.put(`${API_URL}/${id}`, book);
};

const deleteBook = (id) => {
    return axios.delete(`${API_URL}/${id}`);
};


const BookService = {
    getBooks,
    getBook,
    createBook,
    updateBook,
    deleteBook,
    getBooksWithPagination: getBooks // Alias for backward compatibility
};

export default BookService;
